### ISSUE TITLE
Error while compiling libnetconf2

### ISSUE THREAD

**Issue by vprasad** [CESNET/libnetconf2#101]
Hi ,
I am trying to compile the libnetconf2 using qnx 6.2 and gcc 5.2
I am getting the below error :--

[100%] Linking C shared library libnetconf2.so
/opt/QNX652/host/linux/x86/usr/bin/powerpc-unknown-nto-qnx6.5.0-ld: cannot find -lcrypt
/opt/QNX652/host/linux/x86/usr/bin/powerpc-unknown-nto-qnx6.5.0-ld: cannot find -lssh_threads
cc: /opt/QNX652/host/linux/x86/usr/bin/powerpc-unknown-nto-qnx6.5.0-ld error 1
make[2]: *** [CMakeFiles/netconf2.dir/build.make:251: libnetconf2.so.0.11.48] Error 1
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/netconf2.dir/all] Error 2
make: *** [Makefile:130: all] Error 2

Does any one have qnx compiled version of libcrypt and libssh_threads libraries.
If not do u have the source code path of these two libraries so that i can built it .

Thanks,

**Comment by michalvasko** [comment]
Hi,
please use a more recent version of libnetconf2 (`0.12`), that will probably solve your second library problem (`libssh_threads`). As for `libcrypt`, if it is not available for your platform, I cannot help you.

Regards,
Michal

**Comment by vprasad** [comment]
> libcrypt

Hi ,
Thanks for the reply .
Regrading the libcrypt library ,can i comment this from CmakeList and proceed for compilation .
Will this have some major effect on functionalities of libnetconf2 ?
If yes do you have the source code of libcrypt ,so that i can compile it for my platform.?
Thanks,

**Comment by michalvasko** [comment]
> Hi ,
> Thanks for the reply .
> Regrading the libcrypt library ,can i comment this from CmakeList and proceed for compilation .
> Will this have some major effect on functionalities of libnetconf2 ?

Yes, it will fail to compile.

> If yes do you have the source code of libcrypt ,so that i can compile it for my platform.?
> Thanks,

After a quick search I found [this](http://web.mit.edu/freebsd/head/lib/libcrypt/), for example although that is FreeBSD source. However, I do not think it will be as easy as just taking a few source and compiling them into a library, `libcrypt` is most likely not available for your platform by default for a reason.

Regards,
Michal

**Comment by michalvasko** [comment]
Hi,
`crypt(3)` is used to compare SSH-retrieved password with the ones stored on the system, there is no workaround for that. Well, if you do not mind having a modified version of *libnetconf2*, you should still be able to use `publickey` SSH authentication. Also, TLS should work fine except that *netopeer2* will not compile with *libnetconf2* without SSH support.

**Comment by vprasad** [comment]
> Hi,
> `crypt(3)` is used to compare SSH-retrieved password with the ones stored on the system, there is no workaround for that. Well, if you do not mind having a modified version of _libnetconf2_, you should still be able to use `publickey` SSH authentication. Also, TLS should work fine except that _netopeer2_ will not compile with _libnetconf2_ without SSH support.
> 
> Regards,
> Michal

Hi Michal ,
I was able to compile libcrypt.so for qnx and now libnetconf2 is able to link to libcrypt.so .
However still libssh_thread.so is not found.
As per your suggestion ,i upgraded to latest libnetconf2 available in github.But still facing the same issue .Below is the error while doing a make :-  


[100%] Linking C shared library libnetconf2.so
/opt/QNX652/host/linux/x86/usr/bin/powerpc-unknown-nto-qnx6.5.0-ld: cannot find -lssh_threads
cc: /opt/QNX652/host/linux/x86/usr/bin/powerpc-unknown-nto-qnx6.5.0-ld error 1
make[2]: *** [CMakeFiles/netconf2.dir/build.make:251: libnetconf2.so.0.12.26] Error 1
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/netconf2.dir/all] Error 2
make: *** [Makefile:130: all] Error 2

Any suggestions of this ?

Thanks,
Vivek prasad

**Comment by michalvasko** [comment]
Hi,
what version of `libssh` and from what source have you installed?

Regards,
Michal

**Comment by vprasad** [comment]
> Hi,
> what version of `libssh` and from what source have you installed?
> 
> Regards,
> Michal

Hi Michal,

I am using libssh-0.8.3 and i downloaded this code from :-https://www.libssh.org/

Thanks,
Vivek prasad

**Comment by michalvasko** [comment]
Hi Vivek,
I see, so the simplest solution is to just comment out `-lssh_threads` beacuse this library is no longer separate since version 0.8.0. However, we tried to detect this in the `CMakeLists.txt` except it is not working in your case for some reason.

Please, try to download version 0.8.5 and try again. If it is still not detected properly, you will have to investigate it yourself. On line `CMakeLists.txt:192` there is a version test and I would start by adding a `message()` somewhere after this line with the value of `LibSSH_VERSION` variable printed.

Regards,
Michal

**Comment by vprasad** [comment]
Thanks,
I am able to compile libnetconf2 and dependencies successfully on QNX ppc.

**Comment by Arvliu** [comment]
@vprasad  @michalvasko 
Hi,
I am using libssh-0.7.7 to cross compile libnetconf2, but it failed after command [make]:
cd /libnetconf2/build
cmake -DENABLE_TLS=ON -DENABLE_SSH=ON -DCMAKE_INSTALL_PREFIX=/liu_newGcc/libnetconf2_arm -DCMAKE_BUILD_TYPE=Release  .. -DCMAKE_C_COMPILER=arm-linux-gnueabihf-gcc  -DOPENSSL_INCLUDE_DIR=/liu_newGcc/openssl_arm/include -DOPENSSL_CRYPTO_LIBRARY="/liu_newGcc/openssl_arm/lib/libcrypto.so"  -DOPENSSL_SSL_LIBRARIES="/liu_newGcc/openssl_arm/lib/libssl.so"  -DLIBYANG_INCLUDE_DIRS=/liu_newGcc/libyang_arm/include -DLIBYANG_LIBRARIES="/liu_newGcc/libyang_arm/lib/libyang.so" -DOPENSSL_ROOT_DIR=/liu_newGcc/openssl_arm -DLIBSSH_INCLUDE_DIRS=/liu_newGcc/libssh_arm/include -DLIBSSH_LIBRARY_DIR="/liu_newGcc/libssh_arm/lib/libssh.so /liu_newGcc/libssh_arm/lib/libssh_threads.so"

cmake ok
make failed with logs as below:
/usr/linux-devkit/bin/../lib/gcc/arm-linux-gnueabihf/8.3.0/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lssh
/usr/linux-devkit/bin/../lib/gcc/arm-linux-gnueabihf/8.3.0/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lssh_threads

I had tried libssh-0.9.0 before, but libnetconf2 without SSH supprot, then i changed to use libssh-0.7.7, but new error appeared.

**Comment by michalvasko** [comment]
Hi,
maybe someone else can help you specifically but this is simply a problem of your dynamic libraries. If you run `ldconfig -p` (if you have it on your platform, you should) you will see all the libraries you can link against. `libssh` will not be there and that is the problem, you probably need to change some paths so that it can be found.

Regards,
Michal

**Comment by Arvliu** [comment]
@michalvasko 
I've specified libssh with cmake parameters as below, but it seems not work.
-DLIBSSH_INCLUDE_DIRS=/liu_newGcc/libssh_arm/include
-DLIBSSH_LIBRARY_DIR="/liu_newGcc/libssh_arm/lib/libssh.so

**Comment by michalvasko** [comment]
Hi,
maybe I did not make myself clear, this is not a CMake problem but your system linker problem (`ld(1)`). I have provided some basic information how you can check whether you have the problem fixed and how to probably fix the problem. For more details try simply searching, it is not a *libnetconf2* issue.

Regards,
Michal

**Comment by msoodan** [comment]
I have tried compiling, same error of missng crypto library.
Ubuntu 26.04 LTS, Let me know if any recent / existing workaround will solve this error.

root@jenkins:/home/jenci/NetconfTools/libnetconf2/build# cmake ..
-- The C compiler identification is GNU 15.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc - broken
CMake Error at /usr/share/cmake-4.2/Modules/CMakeTestCCompiler.cmake:67 (message):
  The C compiler

    "/usr/bin/cc"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: '/home/jenci/NetconfTools/libnetconf2/build/CMakeFiles/CMakeScratch/TryCompile-KwKtdx'

    Run Build Command(s): /usr/bin/cmake -E env VERBOSE=1 /usr/bin/gmake -f Makefile cmTC_a9775/fast
    /usr/bin/gmake  -f CMakeFiles/cmTC_a9775.dir/build.make CMakeFiles/cmTC_a9775.dir/build
    gmake[1]: Entering directory '/home/jenci/NetconfTools/libnetconf2/build/CMakeFiles/CMakeScratch/TryCompile-KwKtdx'
    Building C object CMakeFiles/cmTC_a9775.dir/testCCompiler.c.o
    /usr/bin/cc    -o CMakeFiles/cmTC_a9775.dir/testCCompiler.c.o -c /home/jenci/NetconfTools/libnetconf2/build/CMakeFiles/CMakeScratch/TryCompile-KwKtdx/testCCompiler.c
    Linking C executable cmTC_a9775
    /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_a9775.dir/link.txt --verbose=1
    /usr/bin/x86_64-linux-gnu-ld.bfd: cannot find -lcrypt: No such file or directory
    collect2: error: ld returned 1 exit status
    /usr/bin/cc -lcrypt  CMakeFiles/cmTC_a9775.dir/testCCompiler.c.o -o cmTC_a9775
    gmake[1]: *** [CMakeFiles/cmTC_a9775.dir/build.make:102: cmTC_a9775] Error 1
    gmake[1]: Leaving directory '/home/jenci/NetconfTools/libnetconf2/build/CMakeFiles/CMakeScratch/TryCompile-KwKtdx'
    gmake: *** [Makefile:134: cmTC_a9775/fast] Error 2





  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:3 (project)

**Comment by michalvasko** [comment]
Probably better to open a new issue, but I have reproduced it on Ubuntu 26.04. The issue is that there is `/usr/lib/x86_64-linux-gnu/libcrypt.so.1` symlink but no `/usr/lib/x86_64-linux-gnu/libcrypt.so` symlink. When I created it, ran `ldconfig`, and rerun `cmake`, *libnetconf2* compiled fine. So I am not sure what the solution is and what is the reasoning behind the decision to not have a symlink to the library without any versions.
